lib/Optimizer/TrainingPreparation.cpp

```diff
@@ -0,0 +1,118 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "glow/Optimizer/TrainingPreparation.h"
+#include "glow/Graph/PlaceholderBindings.h"
+
+namespace glow {
+
+namespace {
+void defaultTensorInitializer(Function *F, Node *node, NodeValue &input,
+                              Tensor *tensor) {
+  switch (node->getKind()) {
+  case Kinded::Kind::ConvolutionNodeKind: {
+    ConvolutionNode *CN = llvm::dyn_cast<ConvolutionNode>(node);
+    if (CN->getFilter() == input) {
+      ShapeNHWC idim = ShapeNHWC(CN->getInput().dims());
+      ShapeHW kdim(CN->getKernels());
+      size_t fanIn = kdim.height * kdim.width * idim.c;
+      tensor->init(Tensor::InitKind::Xavier, fanIn, F->getPRNG());
+    } else if (CN->getBias() == input) {
+      tensor->init(Tensor::InitKind::Broadcast, 0.1, F->getPRNG());
+    }
+  } break;
+  case Kinded::Kind::BatchNormalizationNodeKind: {
+    BatchNormalizationNode *BN = llvm::dyn_cast<BatchNormalizationNode>(node);
+    if (BN->getScale() == input) {
+      tensor->init(glow::Tensor::InitKind::Zero, 0, F->getPRNG());
+    } else if (BN->getBias() == input) {
+      tensor->init(glow::Tensor::InitKind::Broadcast, 0.1, F->getPRNG());
+    } else if (BN->getMean() == input) {
+      tensor->zero();
+    } else if (BN->getVar() == input) {
+      tensor->init(glow::Tensor::InitKind::Broadcast, 1.0, F->getPRNG());
+    }
+  } break;
+  case Kinded::Kind::FullyConnectedNodeKind: {
+    FullyConnectedNode *FCN = llvm::dyn_cast<FullyConnectedNode>(node);
+    if (FCN->getWeights() == input) {
+      auto in = FCN->getInput();
+      tensor->init(glow::Tensor::InitKind::Zero, in.dims()[1], F->getPRNG());
+    } else if (FCN->getBias() == input) {
+      tensor->init(glow::Tensor::InitKind::Broadcast, 0.1, F->getPRNG());
+    }
+  } break;
+  default:
+    break;
+  }
+}
+} // namespace
+
+TensorInitializer getDefaultTensorInitializer() {
+  return defaultTensorInitializer;
+}
+
+void prepareFunctionForTraining(
+    Function *F, PlaceholderBindings &bindings,
+    const std::unordered_set<std::string> &legalOutputNames,
+    TensorInitializer &&initializer) {
+
+  auto &nodes = F->getNodes();
+  // Lookup all nodes, skip Storage types, enumerate inputs,
+  // replace Constant type with trainable Placeholders except special cases,
+  // like BatchNormalization inputs (mean and variance). In special cases
+  // replace Constant type with non-trainable Placeholders.
+  for (auto &node : nodes) {
+    // Skip storages.
+    if (Storage::classof(&node)) {
+      continue;
+    }
+
+    // Skip output node(s).
+    if (legalOutputNames.count(node.getName())) {
```

Comment:
Do we need this information to be explicitly provided? Can we infer which nodes are outputs? I'll admit I didn't think about this too deeply but I think `Placeholders` that are written to only once by `SaveNodes` should be outputs, right?

Reply:
Loader has information about output nodes, but once we have decoupled API from Loader, it's kind of hard to do. I have tried and it looked like a hack with insights into Loader code.

Reply:
For instance, the output node is SoftMax in my testing example, not SaveNode and it can be other types besides SoftMax.

Reply:
Doesn't the SoftMax feed into a Save, though?

Reply:
Let me try, traverse nodes twice, one to find SaveNode and the correspondent input (whatever it is) and skip Placeholder replacement for that.